### ISSUE TITLE
ERROR: piece  missing from gapped contig 

### ISSUE THREAD

**Issue by olekto** [marbl/verkko#234]
Hi,
I get an error when trying to run verkko on a dataset of a fish consisting of ONT, HiFi and Hi-C. In `combineConsensus.err `in 8-hicPipeline/final_contigs/7-consensus I get this error: 
`ERROR:piece  missing from gapped contig haplotype1-0000121. Piece is [N1758N]. numn is <re.Match object; span=(0, 8), match='[N1758N]'>.` 

(I added the last to points to try to trace it down a bit.) haplotype1-0000121 looks like this:
```
path haplotype1-0000121 haplotype1_from_utig4-3261
[N1758N]
[N13652N]
piece001617
[N1000N]
piece001618
[N34607N]
piece001619
[N17605N]
piece001620
end
```
I guess the issue is that this sequence starts with a gap (or two), and no sequence? Is there an easy way to trace down what has happened? Can I just comment this (https://github.com/marbl/verkko/blob/91070e8eb79cf4b7717fc07662b7089e1b7a8336/src/scripts/fasta_combine.py#L92) line and continue? 

Thank you.

Ole

**Comment by skoren** [comment]
I think this is because of the double gap at the start but it's a little strange to have the sequence start with a gap. Can you post what the tsv from your phasing (I'm guessing in 8-hicPipeline) has for that sequence (haplotype1_from_utig4-3261). You can certainly complete the asm by trimming the two gaps at the start of the scfmap to at least get sequences but I'd want to get to the root of why this happened originally.

**Comment by olekto** [comment]
Thank you, Sergey.

You mean from rukki.paths.tsv? This is the content there:
`haplotype1_from_utig4-3261	utig4-50521-,[N13652N:ambig_bubble],utig4-36082-,[N1000N:ambig_bubble],utig4-45789-,utig4-22110-,[N34607N:ambig_bubble],utig4-51215+,[N17605N:ambig_bubble],utig4-17097-,utig4-17099+,utig4-3264-,utig4-3260-,utig4-3261+`

Am I correct in assume that utig4-50521 is lost somewhere? From filtered.gfa in 8-hicPipeline:
`S	utig4-50521	*	LN:i:1758	RC:i:14222	ll:f:8.1`
It is the same length as the first gap (1758).

It does seem to have some coverage. 
`utig4-50521	8.09	1758` in unitig-unrolled-unitig-unrolled-popped-unitig-normal-connected-tip.hifi-coverage.csv
`utig4-50521	5.00	1758` in unitig-unrolled-unitig-unrolled-popped-unitig-normal-connected-tip.ont-coverage.csv

It is not listed in unitig-popped.layout.scfmap.dropped in 8-hicPipeline/final_contigs/6-layoutContigs.

However, in 6-layoutContigs (not inside 8-hicPipeline), it is mention in unitig-popped.layout.scfmap.dropped:
`utig4-50521 has no reads assigned and is not output.`

I hope this helps. I tried tracing it further, but did not get anywhere.

Ole

**Comment by skoren** [comment]
Yeah it seems the short tig didn't get any reads assigned to it and ended up as a gap itself. Might be worth taking a look at the graph near that node at the ambiguous bubble. It's possible you can manually resolve it and update the path and have reads assigned to it without gaps. If you share the GFA (you can remove the sequences by running):

```
cat 5-untip/unitig-unrolled-unitig-unrolled-popped-unitig-normal-connected-tip.gfa |awk '{if (match($1, "^S")) { print $1"\t"$2"\tLN:i:"length($3); } else print $0}' > tmp.gfa
verkko/src/scripts/inject_coverage.py --allow-absent 5-untip/unitig-unrolled-unitig-unrolled-popped-unitig-normal-connected-tip.hifi-coverage.csv tmp.gfa > 5-untip/5-untip/unitig-unrolled-unitig-unrolled-popped-unitig-normal-connected-tip.noseq.gfa
```
this normally gets run after the assembly but this way you can generate the smaller GFA you can upload here to take a look around those nodes. Include the hicverkko.colors.tsv file too when you upload the gfa.

**Comment by olekto** [comment]
Thank you, Sergey. I have uploaded the files here:
[unitig-unrolled-unitig-unrolled-popped-unitig-normal-connected-tip.noseq.gfa.gz](https://github.com/marbl/verkko/files/14506930/unitig-unrolled-unitig-unrolled-popped-unitig-normal-connected-tip.noseq.gfa.gz)
[hicverkko.colors.tsv.gz](https://github.com/marbl/verkko/files/14506941/hicverkko.colors.tsv.gz)

I have to say that this assembly is a bit fragmented. I guess optimally for verkko, both HiFi, but especially, ONT should have had higher coverage. However, it is fun to see how the data performs in verkko. It is a salmon. 

I hope I did the operation correct. I looked at this in BandageNG, but not exactly sure what I should look for.

**Comment by skoren** [comment]
I had a mistake in the command above, missing a * in the output (should have been `print $1"\t"$2"\t*\tLN:i:"length($3);`) so I updated the gfa to fix that and uploaded it. The hifi coverage looks low but passable (20x?) but I'd guess you have very low ONT coverage as almost none of the graph is resolved and it's very fragmented (<10x?). I think the HiC phasing essentially didn't work here. There are lots of paths with same haplotypes on both bubble nodes. Here's the component in question:
<img width="1024" alt="Screenshot 2024-03-06 at 10 15 58 AM" src="https://github.com/marbl/verkko/assets/923478/6f0fe297-6585-446d-a227-1162d1b9c6eb">

It starts from the top left of the figure and the looks reasonable with nodes from both haplotypes but by the end all the nodes are being assigned to a single haplotype even though they are clearly bubble structures. I'd guess this is an issue with the default heterozygosity and length thresholds given the low coverage and fragmented assembly. So you could fill in/extend this path but I wouldn't bother as the phasing in this assembly doesn't look like it'd be accurate and extending this one path isn't going to solve the other places in the graph where the haplotypes look incorrect. @Dmitry-Antipov do you think it is worth investigating what he HiC performed poorly here? @snurk I'm actually also surprised rukki was willing to go through some of these bubbles where both sides are the same haplotype. Is that expected behavior?
[unitig-unrolled-unitig-unrolled-popped-unitig-normal-connected-tip.noseq.gfa.gz](https://github.com/marbl/verkko/files/14512040/unitig-unrolled-unitig-unrolled-popped-unitig-normal-connected-tip.noseq.gfa.gz)

**Comment by skoren** [comment]
I'd update the path to:

```
haplotype1_from_utig4-3261	utig4-50522-, utig4-50521-, utig4-36085-, utig4-36082-, utig4-36084+, utig4-45789-, utig4-22110-, utig4-22111+, utig4-51215+, utig4-17100-, utig4-17097-,utig4-17099+,utig4-3264-,utig4-3260-,utig4-3261+
```
or rather the gaf file to be:
```
haplotype1_from_utig4-3261	<utig4-50522<utig4-50521<utig4-36085<utig4-36082>utig4-36084<utig4-45789<utig4-22110>utig4-22111>utig4-51215<utig4-17100<utig4-17097>utig4-17099<utig4-3264<utig4-3260>utig4-3261
```

Then you can resume the assembly. Depending on how many of these components there are, the N50 of 3-4 mb is possible though like I said, I expect lots of haplotype switches and likely haplotype[12] files will be uneven in size.

**Comment by olekto** [comment]
Super, thank you. 
You meant update the consensus_paths.txt file with the last entry there? It looks like that format at least. I'll report back when it finishes.
